node/node.go

```diff
@@ -869,6 +869,18 @@ func (node *AlgorandFullNode) OnNewBlock(block bookkeeping.Block, delta ledgerco
 	node.hasSyncedSinceStartup = true
 	node.syncStatusMu.Unlock()
 
+	// Look for keyreg events and notify the participationRegistry.
+	for _, tx := range block.Payset {
+		if tx.Txn.Type == protocol.KeyRegistrationTx {
+			id := account.MakeParticipationID(tx.Txn.Sender, tx.Txn.KeyregTxnFields)
+			err := node.participationRegistry.Register(id, block.BlockHeader.Round)
+			// If the key is not installed on this node, ErrParticipationIDNotFound is quickly returned.
+			if err != nil && err != account.ErrParticipationIDNotFound {
+				node.log.Error("Problem with participationRegistry.Register: %w", err)
+			}
+		}
+	}
+
```

Comment:
I'm not sure that this is the best way to do that. Is it possible to conceptually defer this to the moment where we actually need the information and check on the ledger ?

Reply:
So when GetKeys is called it fetches all keys here and checks if any should be registered?

I think that could work and be less error-prone but require more account lookups. I'll investigate.

Reply:
I think that GetKeys would be called only on user-driven cases, whereas this would be called on each and every round.. so we should try and make it so that we push the "extra work" to be per-user case.